File: 88/main.ts
/**
 Do not return anything, modify nums1 in-place instead.
 */
function merge(nums1: number[], m: number, nums2: number[], n: number): void {
  // do nothing
  if (m === 0) {
    // [], [...]
    for (let i = 0; i < n; i++) {
      nums1[i] = nums2[i];
    }
    return;
  }
  if (n === 0) return;

  // from the back so we don't have to move
  let i = m - 1;
  let j = n - 1;
  let current_spot = m + n - 1;

  while (current_spot >= 0) {
    console.log(i, j, current_spot);
    console.log(nums1);
    if (i >= 0 && j >= 0) {
      if (nums1[i] >= nums2[j]) {
        nums1[current_spot] = nums1[i--];
      } else {
        nums1[current_spot] = nums2[j--];
      }
    } else {
      if (i < 0) {
        nums1[current_spot] = nums2[j--];
      } else if (j < 0) {
        nums1[current_spot] = nums1[i--];
      }
    }

    current_spot--;
  }
}

File: 88/main.py
from typing import List


class Solution:
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        """
        * 2 pointers
        * go from back to keep it simple
        """
        target = len(nums1) - 1
        a, b = m - 1, n - 1
        while a >= 0 and b >= 0:
            val1, val2 = nums1[a], nums2[b]
            if val1 < val2:
                # swap b
                nums1[target] = nums2[b]
                target -= 1
                b -= 1
            else:
                nums1[target] = nums1[a]
                target -= 1
                a -= 1
        while b >= 0:
            nums1[target] = nums2[b]
            target -= 1
            b -= 1
